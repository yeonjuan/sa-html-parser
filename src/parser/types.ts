export enum InsertionMode {
  initialMode,
  afterInitialMode,
  // beforeHtmlMode,
  // beforeHeadMode,
  // inHeadMode,
  // inHeadNoScriptMode,
  // afterHeadMode,
  // inBodyMode,
  // textMode,
  // inTableMode,
  // inTableTextMode,
  // inCaptionMode,
  // inColumnGroupMode,
  // inTableBodyMode,
  // inRowMode,
  // inCellMode,
  // inSelectMode,
  // inSelectInTableMode,
  // inTemplateMode,
  // afterBodyMode,
  // inFramesetMode,
  // afterFramesetMode,
  // afterAfterBodyMode,
  // afterAfterFramesetMode,
}
