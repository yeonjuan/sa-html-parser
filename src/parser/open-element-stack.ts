export class OpenElementStack {
  public readonly elements: any[] = [];
  public stackTop = -1;
  public top: any = null;

  public push(element: any) {
    this.top = element;
    this.stackTop++;
    this.elements[this.stackTop] = element;
  }

  public pop() {
    this.stackTop--;
    this.top = this.elements[this.stackTop];
  }

  public popUntilElementPopped(element: any): any[] {
    let elements: any[] = [];
    while (this.stackTop > -1) {
      const poppedElement = this.top;
      elements.push(poppedElement);
      this.pop();
      if (poppedElement === element) {
        break;
      }
    }
    return elements;
  }
}
