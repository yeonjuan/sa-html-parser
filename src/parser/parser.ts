import {
  CommentNode,
  DoctypeNode,
  RootNode,
  AnyNode,
  ElementNode,
  ClosingElementNode,
  TextNode,
} from "../nodes";
import { Tokenizer } from "../tokenizer/tokenizer";
import {
  AnyHtmlToken,
  AttributeToken,
  CharacterLikeToken,
  CommentToken,
  DoctypeToken,
  EndTagToken,
  EofToken,
  HtmlTokenType,
  NullToken,
  StartTagToken,
} from "../tokens";
import { OpenElementStack } from "./open-element-stack";
import * as utils from "../common/utils";

export class Parser {
  private root!: RootNode;
  private html!: string;
  private tokenizer!: Tokenizer;
  private openElementStack = new OpenElementStack();
  private running = true;

  constructor() {}

  public parse(html: string): RootNode {
    this.html = html;
    this.tokenizer = Tokenizer.create(this.html);
    this.createRoot();

    while (this.running) {
      const token = this.tokenizer.getNextToken();
      if (!token) {
        break;
      }
      this.process(token);
    }

    return this.root;
  }

  private createRoot() {
    this.root = new RootNode();
    this.openElementStack.push(this.root);
  }

  private insertCommentNodeToRoot(node: CommentNode) {
    this.insertToCurrent(node);
    this.root.comments.push(node);
  }

  private insertToCurrent(node: AnyNode) {
    this.openElementStack.top.children.push(node);
  }

  private pushToOpenStack(node: any) {
    this.openElementStack.push(node);
  }

  private popFromOpenStackUntilTagName(tagName: string): null | any[] {
    let unclosedElements: null | any[] = null;
    for (let i = this.openElementStack.stackTop; i >= 0; i--) {
      const element = this.openElementStack.elements[i];
      if (element.tagName === tagName) {
        unclosedElements = this.openElementStack.popUntilElementPopped(element);
        break;
      }
    }
    return unclosedElements;
  }

  private process(token: AnyHtmlToken) {
    this[token.type]?.(token as any);
  }

  private [HtmlTokenType.Comment](token: CommentToken) {
    const comment = CommentNode.fromToken(token);
    this.insertToCurrent(comment);
    this.insertCommentNodeToRoot(comment);
  }

  private [HtmlTokenType.Doctype](token: DoctypeToken) {
    this.insertToCurrent(DoctypeNode.fromToken(token));
  }

  private [HtmlTokenType.StartTag](token: StartTagToken) {
    const tagNode = ElementNode.fromToken(token);
    this.insertToCurrent(tagNode);
    if (token.selfClosing) {
      tagNode.selfClosing = true;
    } else {
      this.pushToOpenStack(tagNode);
    }
  }

  private [HtmlTokenType.EndTag](token: EndTagToken) {
    const closing = ClosingElementNode.fromToken(token);
    const poppedElements = this.popFromOpenStackUntilTagName(
      token.tagName.value
    );
    if (!poppedElements) {
      throw new Error();
    }
    const element = utils.last<ElementNode>(poppedElements)!;
    element.children = utils.getChildrenRecursively(element);
    element.closingElement = closing;
    element.end = closing.end;
    element.loc.end = closing.loc.end;
    element.range[1] = closing.range[1];
  }

  private [HtmlTokenType.CharacterLike](token: CharacterLikeToken) {
    if (this.openElementStack.top.children?.length) {
      const lastChild: any = utils.last(this.openElementStack.top.children);
      if (lastChild.type === "Text") {
        const textNode = lastChild as TextNode;
        textNode.value += token.value.value;
        textNode.end = token.end;
        textNode.loc.end = token.value.loc.end;
        return;
      }
    }
    this.insertToCurrent(TextNode.fromToken(token));
  }

  // eslint-disable-next-line @typescript-eslint/no-unused-vars
  private [HtmlTokenType.EOF](token: EofToken) {
    const poppedElements = this.openElementStack.popUntilBeforeElementPopped(
      this.root
    );
    if (poppedElements.length) {
      for (let i = poppedElements.length - 2; i >= 0; i--) {
        this.root.children.push(poppedElements[i]);
      }
    }
    this.running = false;
    const lastChild = utils.last<AnyNode>(this.root.children);
    this.root.range[1] = lastChild?.range[1] || 0;
    this.root.end = lastChild?.range[1] || 0;
    this.root.loc.end = lastChild?.loc.end || {
      column: 0,
      line: 1,
    };
  }

  // eslint-disable-next-line @typescript-eslint/no-unused-vars
  private [HtmlTokenType.Null](token: NullToken) {}
  // eslint-disable-next-line @typescript-eslint/no-unused-vars
  private [HtmlTokenType.Attribute](token: AttributeToken) {}
}
