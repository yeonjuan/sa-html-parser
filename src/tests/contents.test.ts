import { parse } from "../parser";
import * as fs from "fs";
import * as path from "path";

describe("contents", () => {
  const contentsHTML = fs.readFileSync(
    path.join(__dirname, "contents-html"),
    "utf-8"
  );

  it("contents.html", () => {
    const result = parse(contentsHTML);
    expect(result).toMatchSnapshot();
  });
});
