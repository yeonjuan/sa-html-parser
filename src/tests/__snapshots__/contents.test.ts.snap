// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`contents contents.html 1`] = `
Root {
  "children": Array [
    Doctype {
      "end": 15,
      "loc": Object {
        "end": Object {
          "column": 15,
          "line": 1,
        },
        "start": Object {
          "column": 0,
          "line": 1,
        },
      },
      "name": DoctypeName {
        "end": 14,
        "loc": Object {
          "end": Object {
            "column": 14,
            "line": 1,
          },
          "start": Object {
            "column": 10,
            "line": 1,
          },
        },
        "range": Array [
          10,
          14,
        ],
        "start": 10,
        "type": "DoctypeName",
        "value": "html",
      },
      "publicId": null,
      "range": Array [
        0,
        15,
      ],
      "start": 0,
      "systemId": null,
      "type": "Doctype",
    },
    Text {
      "end": 16,
      "loc": Object {
        "end": Object {
          "column": 16,
          "line": 1,
        },
        "start": Object {
          "column": 15,
          "line": 1,
        },
      },
      "range": Array [
        15,
        16,
      ],
      "start": 15,
      "type": "Text",
      "value": "
",
    },
    Element {
      "children": Array [
        Text {
          "end": 25,
          "loc": Object {
            "end": Object {
              "column": 2,
              "line": 3,
            },
            "start": Object {
              "column": 6,
              "line": 2,
            },
          },
          "range": Array [
            22,
            25,
          ],
          "start": 22,
          "type": "Text",
          "value": "
  ",
        },
        Element {
          "children": Array [
            Text {
              "end": 36,
              "loc": Object {
                "end": Object {
                  "column": 4,
                  "line": 4,
                },
                "start": Object {
                  "column": 8,
                  "line": 3,
                },
              },
              "range": Array [
                31,
                36,
              ],
              "start": 31,
              "type": "Text",
              "value": "
    ",
            },
            Element {
              "children": Array [
                Text {
                  "end": 44,
                  "loc": Object {
                    "end": Object {
                      "column": 12,
                      "line": 4,
                    },
                    "start": Object {
                      "column": 9,
                      "line": 4,
                    },
                  },
                  "range": Array [
                    41,
                    44,
                  ],
                  "start": 41,
                  "type": "Text",
                  "value": "1<2",
                },
              ],
              "closingElement": ClosingElement {
                "end": 50,
                "loc": Object {
                  "end": Object {
                    "column": 18,
                    "line": 4,
                  },
                  "start": Object {
                    "column": 12,
                    "line": 4,
                  },
                },
                "name": ElementName {
                  "end": 49,
                  "loc": Object {
                    "end": Object {
                      "column": 17,
                      "line": 4,
                    },
                    "start": Object {
                      "column": 14,
                      "line": 4,
                    },
                  },
                  "range": Array [
                    46,
                    49,
                  ],
                  "start": 46,
                  "type": "ElementName",
                  "value": "div",
                },
                "range": Array [
                  44,
                  50,
                ],
                "start": 44,
                "type": "ClosingElement",
              },
              "end": 50,
              "loc": Object {
                "end": Object {
                  "column": 18,
                  "line": 4,
                },
                "start": Object {
                  "column": 4,
                  "line": 4,
                },
              },
              "openingElement": OpeningElement {
                "attributes": Array [],
                "end": 41,
                "loc": Object {
                  "end": Object {
                    "column": 9,
                    "line": 4,
                  },
                  "start": Object {
                    "column": 4,
                    "line": 4,
                  },
                },
                "name": ElementName {
                  "end": 40,
                  "loc": Object {
                    "end": Object {
                      "column": 8,
                      "line": 4,
                    },
                    "start": Object {
                      "column": 5,
                      "line": 4,
                    },
                  },
                  "range": Array [
                    37,
                    40,
                  ],
                  "start": 37,
                  "type": "ElementName",
                  "value": "div",
                },
                "range": Array [
                  36,
                  41,
                ],
                "selfClosing": false,
                "start": 36,
                "type": "OpeningElement",
              },
              "range": Array [
                36,
                50,
              ],
              "start": 36,
              "type": "Element",
            },
            Text {
              "end": 53,
              "loc": Object {
                "end": Object {
                  "column": 2,
                  "line": 5,
                },
                "start": Object {
                  "column": 18,
                  "line": 4,
                },
              },
              "range": Array [
                50,
                53,
              ],
              "start": 50,
              "type": "Text",
              "value": "
  ",
            },
          ],
          "closingElement": ClosingElement {
            "end": 60,
            "loc": Object {
              "end": Object {
                "column": 9,
                "line": 5,
              },
              "start": Object {
                "column": 2,
                "line": 5,
              },
            },
            "name": ElementName {
              "end": 59,
              "loc": Object {
                "end": Object {
                  "column": 8,
                  "line": 5,
                },
                "start": Object {
                  "column": 4,
                  "line": 5,
                },
              },
              "range": Array [
                55,
                59,
              ],
              "start": 55,
              "type": "ElementName",
              "value": "body",
            },
            "range": Array [
              53,
              60,
            ],
            "start": 53,
            "type": "ClosingElement",
          },
          "end": 60,
          "loc": Object {
            "end": Object {
              "column": 9,
              "line": 5,
            },
            "start": Object {
              "column": 2,
              "line": 3,
            },
          },
          "openingElement": OpeningElement {
            "attributes": Array [],
            "end": 31,
            "loc": Object {
              "end": Object {
                "column": 8,
                "line": 3,
              },
              "start": Object {
                "column": 2,
                "line": 3,
              },
            },
            "name": ElementName {
              "end": 30,
              "loc": Object {
                "end": Object {
                  "column": 7,
                  "line": 3,
                },
                "start": Object {
                  "column": 3,
                  "line": 3,
                },
              },
              "range": Array [
                26,
                30,
              ],
              "start": 26,
              "type": "ElementName",
              "value": "body",
            },
            "range": Array [
              25,
              31,
            ],
            "selfClosing": false,
            "start": 25,
            "type": "OpeningElement",
          },
          "range": Array [
            25,
            60,
          ],
          "start": 25,
          "type": "Element",
        },
        Text {
          "end": 61,
          "loc": Object {
            "end": Object {
              "column": 10,
              "line": 5,
            },
            "start": Object {
              "column": 9,
              "line": 5,
            },
          },
          "range": Array [
            60,
            61,
          ],
          "start": 60,
          "type": "Text",
          "value": "
",
        },
      ],
      "closingElement": ClosingElement {
        "end": 68,
        "loc": Object {
          "end": Object {
            "column": 7,
            "line": 6,
          },
          "start": Object {
            "column": 0,
            "line": 6,
          },
        },
        "name": ElementName {
          "end": 67,
          "loc": Object {
            "end": Object {
              "column": 6,
              "line": 6,
            },
            "start": Object {
              "column": 2,
              "line": 6,
            },
          },
          "range": Array [
            63,
            67,
          ],
          "start": 63,
          "type": "ElementName",
          "value": "html",
        },
        "range": Array [
          61,
          68,
        ],
        "start": 61,
        "type": "ClosingElement",
      },
      "end": 68,
      "loc": Object {
        "end": Object {
          "column": 7,
          "line": 6,
        },
        "start": Object {
          "column": 0,
          "line": 2,
        },
      },
      "openingElement": OpeningElement {
        "attributes": Array [],
        "end": 22,
        "loc": Object {
          "end": Object {
            "column": 6,
            "line": 2,
          },
          "start": Object {
            "column": 0,
            "line": 2,
          },
        },
        "name": ElementName {
          "end": 21,
          "loc": Object {
            "end": Object {
              "column": 5,
              "line": 2,
            },
            "start": Object {
              "column": 1,
              "line": 2,
            },
          },
          "range": Array [
            17,
            21,
          ],
          "start": 17,
          "type": "ElementName",
          "value": "html",
        },
        "range": Array [
          16,
          22,
        ],
        "selfClosing": false,
        "start": 16,
        "type": "OpeningElement",
      },
      "range": Array [
        16,
        68,
      ],
      "start": 16,
      "type": "Element",
    },
    Text {
      "end": 69,
      "loc": Object {
        "end": Object {
          "column": 8,
          "line": 6,
        },
        "start": Object {
          "column": 7,
          "line": 6,
        },
      },
      "range": Array [
        68,
        69,
      ],
      "start": 68,
      "type": "Text",
      "value": "
",
    },
  ],
  "comments": Array [],
  "end": 69,
  "errors": Array [
    ParsingError {
      "column": 11,
      "error": "invalid-first-character-of-tag-name",
      "index": 43,
      "line": 4,
    },
  ],
  "loc": Object {
    "end": Object {
      "column": 8,
      "line": 6,
    },
    "start": Object {
      "column": 0,
      "line": 1,
    },
  },
  "range": Array [
    0,
    69,
  ],
  "start": 0,
  "tokens": Array [],
  "type": "Root",
}
`;
